tensorboard/loader.py

```diff
@@ -732,85 +740,367 @@ def get_next_event(self):
+    self._update_first_step_invocations += 1
+    # proto3 doesn't let us distinguish between absent and 0.
+    if event.step:
+      step = 0 if self._observed_zero_step else event.step
+      if self._first_step is None or step < self._first_step:
+        self._first_step = step
+    elif event.summary:
+      self._observed_zero_step = True
+
+  def __hash__(self):
+    return hash(self._key)
+
+  def __eq__(self, other):
+    return self._key == other._key
+
+  def __lt__(self, other):
+    return self._key < other._key
+
+  def __str__(self):
+    offset = self.get_offset()
+    if offset:
+      return u'EventLogReader{path=%s, offset=%d}' % (self.path, offset)
+    else:
+      return u'EventLogReader{%s}' % self.path
+
+
+@util.closeable
+@functools.total_ordering
+@six.python_2_unicode_compatible
+class RunReader(object):
+  """Utility for loading event logs into the DB.
+
+  This class merges the chain of event log files into one meaningful
+  stream of events, ordered by step or timestamp.
+
+  Fields:
+    rowid: The primary key of the corresponding row in Runs.
+    name: Display name of this run.
+  """
+
+  def __init__(self, db_conn, rowid, name):
+    """Creates new instance.
+
+    Args:
+      db_conn: A PEP 249 Connection object.
+      rowid: Primary key of run in `Runs` table, which should already
+          be inserted. This is a bit-packed int made by db.RUN_ROWID.
+      run_id: The run_id column which is encoded in the rowid bits.
+      name: Display name of run.
+
+    :type db_conn: db.Connection
+    :type rowid: int
+    :type name: str
+    """
+    self._db_conn = db_conn
+    self.rowid = db.RUN_ROWID.check(rowid)
+    self.run_id = db.RUN_ROWID.parse(rowid)[1]
+    self.name = tf.compat.as_text(name)
+    self._mark = -1
+    self._logs = []  # type: list[EventLogReader]
+    self._i = 0
+    self._entombed_progress = 0
+    self._has_new_stuff = False
+    self._saved_events = \
+        collections.deque()  # type: collections.deque[tf.Event]
+    self._prepended_events = \
+        collections.deque()  # type: collections.deque[tf.Event]
+
+  def add_event_log(self, log):
+    """Adds event log to run loader.
+
+    Event logs must be added monotonically, based on the timestamp in
+    the filename. Please note that calling this method could cause a
+    current batch of reads to fast forward.
+
+    Args:
+      log: An EventLogReader instance.
+
+    Returns:
+      True if log was actually added.
+
+    :type log: EventLogReader
+    :rtype: bool
+    """
+    if self._logs and log <= self._logs[-1]:
+      return False
+    with contextlib.closing(self._db_conn.cursor()) as c:
+      c.execute(
+          'SELECT rowid, offset FROM EventLogs WHERE run_id = ? AND path = ?',
+          (self.run_id, log.path))
+      row = c.fetchone()
+      if row:
+        log.rowid = row[0]
+        log.set_offset(row[1])
+      else:
+        event_log_id = db.EVENT_LOG_ID.generate()
```

Comment:
How so? `TensorBase.run_transaction` takes a `Connection -> T`, but `RunReader` needs to be given a `db_conn` at construction time. How does the transaction-enabled connection get to `RunReader`? (Could you show a simple example?)